autogpt_platform/frontend/src/components/CustomNode.tsx

```diff
@@ -92,812 +92,826 @@ export type CustomNodeData = {
 
 export type CustomNode = XYNode<CustomNodeData, "custom">;
 
-export function CustomNode({
-  data,
-  id,
-  width,
-  height,
-  selected,
-}: NodeProps<CustomNode>) {
-  const [isOutputOpen, setIsOutputOpen] = useState(data.isOutputOpen || false);
-  const [isAdvancedOpen, setIsAdvancedOpen] = useState(false);
-  const [isModalOpen, setIsModalOpen] = useState(false);
-  const [activeKey, setActiveKey] = useState<string | null>(null);
-  const [inputModalValue, setInputModalValue] = useState<string>("");
-  const [isOutputModalOpen, setIsOutputModalOpen] = useState(false);
-  const { updateNodeData, deleteElements, addNodes, getNode } = useReactFlow<
-    CustomNode,
-    Edge
-  >();
-  const isInitialSetup = useRef(true);
-  const flowContext = useContext(FlowContext);
-  const api = useBackendAPI();
-  const { formatCredits } = useCredits();
-  let nodeFlowId = "";
-
-  if (data.uiType === BlockUIType.AGENT) {
-    // Display the graph's schema instead AgentExecutorBlock's schema.
-    data.inputSchema = data.hardcodedValues?.input_schema || {};
-    data.outputSchema = data.hardcodedValues?.output_schema || {};
-    nodeFlowId = data.hardcodedValues?.graph_id || nodeFlowId;
-  }
-
-  if (!flowContext) {
-    throw new Error("FlowContext consumer must be inside FlowEditor component");
-  }
-
-  const { setIsAnyModalOpen, getNextNodeId } = flowContext;
-
-  useEffect(() => {
-    if (data.executionResults || data.status) {
-      setIsOutputOpen(true);
+export const CustomNode = React.memo(
+  function CustomNode({
```

Comment:
This wouldn't cause such a massive diff and likely merge conflicts:
```suggestion
export const CustomNode = React.memo(CustomNodeInternal);

function CustomNodeInternal({
```

Reply:
I don't know, I know my version looks worse but this is unnecessary diff and nesting.

Reply:
> Even after removing hover states, CustomNode can still be expensive if it re-renders frequently. You can wrap it with React.memo (or useMemo around large sub-sections) so that it only re-renders when its props.data really changes

This is the advice from o1, we can remove it if you know its not useful